pyfar/classes/audio.py

```diff
@@ -1021,7 +1039,122 @@ def power(data: tuple, domain='freq'):
     return _arithmetic(data, domain, _power)
 
 
-def _arithmetic(data: tuple, domain: str, operation: Callable):
+def matrix_multiplication(
+        data: tuple, domain='freq', axes=[(-2, -1), (-2, -1), (-2, -1)]):
+    """Matrix multiplication of multidimensional pyfar audio objects and/or
+    array likes.
+
+    This function is based on ``numpy.matmul`` and only acts on the channels,
+    not on the frequency/time dimension. Pyfar audio objects are:
+    :py:func:`Signal`, :py:func:`TimeData`, and :py:func:`FrequencyData`.
+
+    Parameters
+    ----------
+    data : tuple of the form (data_1, data_2, ..., data_N)
+        Data to be multiplied. Can contain pyfar audio objects or array likes.
+        The data is `broadcasted
+        <https://numpy.org/doc/stable/user/basics.broadcasting.html>`_
+        to the data with the highest number of channel dimensions. Pyfar audio
+        objects can not be mixed, e.g., :py:func:`TimeData` and
+        :py:func:`FrequencyData` objects do not work together. See
+        :py:mod:`arithmetic operations <pyfar._concepts.arithmetic_operations>`
+        for possible combinations of Signal FFT normalizations.
+    domain : ``'time'``, ``'freq'``, optional
+        Flag to indicate if the operation should be performed in the time or
+        frequency domain. If working in the frequency domain, the FFT
+        normalization is removed before the operation (See
+        :py:func:`~pyfar.dsp.fft.normalization`). The default is ``'freq'``.
+    axes : list of 3 tuples
+        A list of tuples with the indices of the channels the
+        multiplication should operate on. The signature is (i,j),(j,k)->(i,k),
+        so the base elements of the multiplication are two-dimensional
+        matrices along the time/frequency dimension taken to be stored in the
+        indicated channels of the result. Negative values in ``axes`` refer to
+        ``cshape``, as the frequency/time axis is excluded. For instance, to
+        multiply matrices residing (stacked) in the last two channel
+        dimensions to be stored in the last two channel dimensions, this is
+        indicated by ``[(-2, -1), (-2, -1), (-2, -1)]``, which is the default.
+        See notes and examples below.
+
+    Returns
+    -------
+    results : Signal, TimeData, FrequencyData, numpy array
+        Result of the operation as numpy array, if `data` contains only array
+        likes and numbers. Result as pyfar audio object if `data` contains an
+        audio object. The `fft_norm` is ``'none'`` if all FFT norms are
+        ``'none'``. Otherwise the first `fft_norm` that is not ``'none'`` is
+        used.
```

Comment:
Is this still intended, or should it use our new 'broadcasting' rules for FFT normalization?

Reply:
This is in line with the docstrings of the other arithmetic operations. What do you refer to?

Reply:
I was thinking about this: https://pyfar.readthedocs.io/en/latest/concepts/pyfar.arithmetic_operations.html. For example, we do not allow two signals with norms 'amplitude' and 'power' to be multiplied. Are the docstrings of the arithmetic functions outdated?

Reply:
Fix according to #320